Assets/MixedRealityToolkit-SDK/Features/Teleportation/MixedRealityTeleportManager.cs

```diff
@@ -0,0 +1,263 @@
+                if (isProcessingTeleportRequest)
+                {
+                    Debug.LogWarning("Couldn't change teleport duration. Teleport in progress.");
+                    return;
+                }
+
+                teleportDuration = value;
+            }
+        }
+
+        private float teleportDuration = 0.25f;
+
+        private static readonly ExecuteEvents.EventFunction<IMixedRealityTeleportHandler> OnTeleportRequestHandler =
+            delegate (IMixedRealityTeleportHandler handler, BaseEventData eventData)
+            {
+                var casted = ExecuteEvents.ValidateEventData<TeleportEventData>(eventData);
+                handler.OnTeleportRequest(casted);
+            };
+
+        /// <inheritdoc />
+        public void RaiseTeleportRequest(IMixedRealityPointer pointer, IMixedRealityTeleportHotSpot hotSpot)
+        {
+            // initialize event
+            teleportEventData.Initialize(pointer, hotSpot);
+
+            // Pass handler
+            HandleEvent(teleportEventData, OnTeleportRequestHandler);
+        }
+
+        private static readonly ExecuteEvents.EventFunction<IMixedRealityTeleportHandler> OnTeleportStartedHandler =
+            delegate (IMixedRealityTeleportHandler handler, BaseEventData eventData)
+            {
+                var casted = ExecuteEvents.ValidateEventData<TeleportEventData>(eventData);
+                handler.OnTeleportStarted(casted);
+            };
+
+        /// <inheritdoc />
+        public void RaiseTeleportStarted(IMixedRealityPointer pointer, IMixedRealityTeleportHotSpot hotSpot)
+        {
+            if (isTeleporting)
+            {
+                Debug.LogError("Teleportation already in progress");
+                return;
+            }
+
+            isTeleporting = true;
+
+            // initialize event
+            teleportEventData.Initialize(pointer, hotSpot);
+
+            // Pass handler
+            HandleEvent(teleportEventData, OnTeleportStartedHandler);
+
+            ProcessTeleportationRequest(teleportEventData);
+        }
+
+        private static readonly ExecuteEvents.EventFunction<IMixedRealityTeleportHandler> OnTeleportCompletedHandler =
+            delegate (IMixedRealityTeleportHandler handler, BaseEventData eventData)
+            {
+                var casted = ExecuteEvents.ValidateEventData<TeleportEventData>(eventData);
+                handler.OnTeleportCompleted(casted);
+            };
+
+        /// <inheritdoc />
+        public void RaiseTeleportComplete(IMixedRealityPointer pointer, IMixedRealityTeleportHotSpot hotSpot)
+        {
+            // Check to make sure no one from outside the Teleport System called this method.
+            // Other implementations may have a different way of processing requests.
+            if (isProcessingTeleportRequest)
+            {
+                Debug.LogError("Calls to this method from outside the Teleport System is not allowed in this implementation.");
```

Comment:
Error message does not appear to actually describe the situation.

what sets/clears isProcessingTeleportRequest?

Reply:
also, there isn't a similar check in RaiseTeleportRequest or the other related events

Reply:
This may be what you are looking for: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/caller-information

Reply:
In our specific implementation, we're handling the complete event ourselves from within the teleport manager. Other implementations may choose to do it in other ways.

Reply:
it still feels like we are not doing what the comment advertises... I am ok with cleaning this up in beta

Reply:
#2557

Reply:
I don't think I understand, but okay, let's do that.

Reply:
> also, there isn't a similar check in RaiseTeleportRequest or the other related events

That's because it's expected that they come from Teleport Pointers.

Reply:
In our specific implementation of they system, the teleport manager is the one telling everyone listening that the teleport has completed. (line 260)

Other people using different implementations may choose to do it differently.

We have to satisfy the interface implementation of `RaiseTeleportCompleted`, and so too do other implementations.

Reply:
I'm going to need to understand the intent / system better. Either way, the work item to change this is scheduled for beta :)

Reply:
I ended up just making RaiseTeleportCompleted private and removed it from the interface.